internal/containerupdate/container_updater.go

```diff
@@ -2,13 +2,71 @@ package containerupdate
 
 import (
 	"context"
+	"fmt"
 	"io"
 
+	"github.com/windmilleng/tilt/internal/container"
+	"github.com/windmilleng/tilt/internal/docker"
+	"github.com/windmilleng/tilt/internal/k8s"
+	"github.com/windmilleng/tilt/internal/mode"
 	"github.com/windmilleng/tilt/internal/model"
 	"github.com/windmilleng/tilt/internal/store"
+	"github.com/windmilleng/tilt/internal/target"
 )
 
 type ContainerUpdater interface {
 	UpdateContainer(ctx context.Context, deployInfo store.DeployInfo,
 		archiveToCopy io.Reader, filesToDelete []string, cmds []model.Cmd, hotReload bool) error
+	SupportsSpecs(specs []model.TargetSpec) (supported bool, msg string)
+}
+
+// Differentiate between the CU we'll need for k8s updates (may be docker || synclet || exec)
+// and the one we'll need for DC updates (always docker)
+type K8sContainerUpdater ContainerUpdater
+type DCContainerUpdater ContainerUpdater
+
+func ProvideK8sContainerUpdater(kCli k8s.Client, dCli docker.Client, sm SyncletManager, env k8s.Env, updMode mode.UpdateMode, runtime container.Runtime) K8sContainerUpdater {
+	if updMode == mode.UpdateModeImage || updMode == mode.UpdateModeNaive {
+		return NewExplodingContainerUpdater()
```

Comment:
i worry that this ExplodingContainerUpdater is a code smell that our abstractions are wrong